src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1096,73 +1108,101 @@ private async ValueTask<(Socket?, Stream?, TransportContext?, HttpResponseMessag
+                            return (null, null, response);
                         }
                         break;
                 }
 
-                Socket? socket = (stream as NetworkStream)?.Socket;
+                Debug.Assert(connection != null);
 
                 TransportContext? transportContext = null;
                 if (_kind == HttpConnectionKind.Https || _kind == HttpConnectionKind.SslProxyTunnel)
                 {
-                    SslStream sslStream = await ConnectHelper.EstablishSslConnectionAsync(allowHttp2 ? _sslOptionsHttp2! : _sslOptionsHttp11!, request, async, stream!, cancellationToken).ConfigureAwait(false);
-                    stream = sslStream;
+                    SslStream sslStream = await ConnectHelper.EstablishSslConnectionAsync(allowHttp2 ? _sslOptionsHttp2! : _sslOptionsHttp11!, request, async, connection.Stream, cancellationToken).ConfigureAwait(false);
+                    connection = Connection.FromStream(sslStream, leaveOpen: false, connection.ConnectionProperties, connection.LocalEndPoint, connection.RemoteEndPoint);
                     transportContext = sslStream.TransportContext;
                 }
 
-                return (socket, stream, transportContext, null);
+                return (connection, transportContext, null);
             }
             finally
             {
                 cancellationWithConnectTimeout?.Dispose();
             }
         }
 
+        private ValueTask<Connection> ConnectToTcpHostAsync(string host, int port, HttpRequestMessage initialRequest, bool async, CancellationToken cancellationToken)
+        {
+            if (async)
+            {
+                ConnectionFactory connectionFactory = Settings._connectionFactory ?? SocketsHttpConnectionFactory.Default;
+
+                var endPoint = new DnsEndPointWithProperties(host, port, initialRequest);
+                return ConnectHelper.ConnectAsync(connectionFactory, endPoint, endPoint, cancellationToken);
+            }
+
+            // Synchronous path.
+
+            if (Settings._connectionFactory != null)
+            {
+                // connection factories only support async.
+                throw new HttpRequestException();
+            }
+
+            try
+            {
+                return new ValueTask<Connection>(ConnectHelper.Connect(host, port, cancellationToken));
+            }
+            catch (Exception ex)
+            {
+                return ValueTask.FromException<Connection>(ex);
+            }
+        }
+
         internal async ValueTask<(HttpConnection?, HttpResponseMessage?)> CreateHttp11ConnectionAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
         {
-            (Socket? socket, Stream? stream, TransportContext? transportContext, HttpResponseMessage? failureResponse) =
+            (Connection? connection, TransportContext? transportContext, HttpResponseMessage? failureResponse) =
```

Comment:
With all this new connection related infrastructure, I have to wonder why `transportContext` isn't just a connection property or something like that.

Reply:
I'm pretty sure it's just taken from the `SslStream`, so this change would be trivial.

Reply:
It is, yeah.